cirq/ops/controlled_operation.py

```diff
@@ -0,0 +1,97 @@
+# Copyright 2019 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import Union, Any
+
+import numpy as np
+
+from cirq import protocols, linalg, value
+from cirq.ops import raw_types
+from cirq.type_workarounds import NotImplementedType
+
+
+@value.value_equality
+class ControlledOperation(raw_types.Operation):
+    def __init__(self,
+                 control: raw_types.QubitId,
+                 sub_operation: raw_types.Operation):
+        self.control = control
+        self.sub_operation = sub_operation
+
+
+    @property
+    def qubits(self):
+        return (self.control,) + self.sub_operation.qubits
+
+    def with_qubits(self, *new_qubits):
+        return ControlledOperation(
+            new_qubits[0],
+            self.sub_operation.with_qubits(*new_qubits[1:]))
+
+    def _decompose_(self):
+        result = protocols.decompose_once_with_qubits(self.sub_operation,
+                                                      self.qubits[1:],
+                                                      NotImplemented)
+        if result is NotImplemented:
+            return NotImplemented
+
+        return [ControlledOperation(self.control, op) for op in result]
+
+    def _value_equality_values_(self):
+        return self.control, self.sub_operation
+
+    def _has_unitary_(self) -> bool:
+        return protocols.has_unitary(self.sub_operation)
+
+    def _unitary_(self) -> Union[np.ndarray, NotImplementedType]:
+        sub_matrix = protocols.unitary(self.sub_operation, None)
+        if sub_matrix is None:
+            return NotImplemented
+        return linalg.block_diag(np.eye(sub_matrix.shape[0]), sub_matrix)
+
+    def __str__(self):
+        return 'C{}'.format(self.control) + str(self.sub_operation)
```

Comment:
This `__str__` will look odd when using named qubits e.g `CaX(b, c)`. I would do `"C{}({})".format(self.sub_operation, ", ".join(str(q) for q in self.qubits))`, matching what you see when you use a controlled gate.

Reply:
ack